lib/rust/profiler-macros/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+fn enabled_profiler(snake_ident: &syn::Ident, ObjIdent: &syn::Ident) -> proc_macro::TokenStream {
+    let start = quote::format_ident!("start_{}", snake_ident);
+    let with_same_start = quote::format_ident!("{}_with_same_start", snake_ident);
+    let ts = quote::quote! {
+        // =================================
+        // === Profiler (e.g. Objective) ===
+        // =================================
+
+        #[derive(Copy, Clone, Debug)]
+        pub struct #ObjIdent(ProfilerId);
+
+
+        // === Trait Implementations ===
+
+        impl Profiler for #ObjIdent {
+            type Started = ProfilerData;
+            fn finish(self, data: &Self::Started) {
+                data.finish(self.0);
+            }
+        }
+
+
+        // === Constructor macros ===
+
+        #[macro_export]
+        macro_rules! #start {
+            ($parent: expr, $label: expr) => {{
+                use profiler::Parent;
+                let label = concat!($label, " (", file!(), ":", line!(), ")");
+                let profiler: profiler::Started<profiler::#ObjIdent> = $parent.new_child(label);
+                profiler
+            }}
+        }
+
+        #[macro_export]
+        macro_rules! #with_same_start {
+            ($parent: expr, $label: expr) => {{
+                use profiler::Parent;
+                let label = concat!($label, " (", file!(), ":", line!(), ")");
+                let profiler: profiler::Started<profiler::#ObjIdent> =
+                    $parent.new_child_same_start(label);
+                profiler
+            }}
+        }
+    };
+    ts.into()
+}
+
+/// Define a profiler type for a disabled profiling level.
+#[allow(non_snake_case)]
+fn disabled_profiler(snake_ident: &syn::Ident, ObjIdent: &syn::Ident) -> proc_macro::TokenStream {
+    let start = quote::format_ident!("start_{}", snake_ident);
+    let with_same_start = quote::format_ident!("{}_with_same_start", snake_ident);
+    let ts = quote::quote! {
+        // =================================
+        // === Profiler (e.g. Objective) ===
+        // =================================
+
+        #[derive(Copy, Clone, Debug)]
+        pub struct #ObjIdent(());
+
+
+        // === Trait Implementations ===
+
+        impl Profiler for #ObjIdent {
+            type Started = ();
+            fn finish(self, _: &Self::Started) {}
+        }
+
+
+        // === Constructor macros ===
+
+        #[macro_export]
+        macro_rules! #start {
+            ($parent: expr, $label: expr) => {
+                profiler::Started { profiler: profiler::#ObjIdent(()), data: () }
+            }
+        }
+        #[macro_export]
+        macro_rules! #with_same_start {
+            ($parent: expr, $label: expr) => {
+                profiler::Started { profiler: profiler::#ObjIdent(()), data: () }
+            }
+        }
+    };
+    ts.into()
+}
+
+/// impl Parent<T> for U, when U is a profiler of an enabled level.
```

Comment:
Please make sure that all comments are valid English sentences starting with a capital letter and describing what the thing does. I understand what you mean here, but describing that a little bit more could help readers to easily read comments.